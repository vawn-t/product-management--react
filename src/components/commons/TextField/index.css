.field-wrapper {
  padding: 1rem 1rem;
  width: 100%;
}

.input-wrapper {
  display: flex;
  position: relative;
  align-items: center;
  flex-direction: row;
  padding: 0.25rem 0;
}

.field {
  border-radius: 3px;
  font-size: 16px;
  width: 100%;
  outline: none;
  padding: 0.5rem;
  text-indent: 0.5rem;
}

/* - Disabled field */
/* - ---------------------------------------------- */
.field:disabled {
  background-color: -internal-light-dark(
    rgba(239, 239, 239, 0.3),
    rgba(59, 59, 59, 0.3)
  );
  color: -internal-light-dark(rgb(84, 84, 84), rgb(170, 170, 170));
}

/* - Standard field */
/* - ---------------------------------------------- */
.field-standard {
  background: var(--light-gray);
  border: 1px solid var(--light-gray);
}

.field-standard:hover {
  border: 1px solid var(--light-gray);
}

/* - Outline field */
/* - ---------------------------------------------- */
.field-outline {
  background: var(--white);
  border: 1px solid var(--light-gray);
}

.field-outline:hover,
.field-outline:focus {
  border-color: var(--dark);
}

/* - Error field */
/* - ---------------------------------------------- */
.field-error {
  background: var(--white);
  border: 1px solid var(--danger);
}

.icon-left {
  position: absolute;
  right: 1.5rem;
}
