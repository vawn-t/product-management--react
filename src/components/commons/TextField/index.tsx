// Libraries
import React, {
  memo,
  forwardRef,
  useState,
  ChangeEvent,
  useCallback
} from 'react';

// Components
import { Image } from '@components/index';

// Styles
import './index.css';

// Types
import {
  ImageVariants,
  TextFieldVariants,
  TypeVariables
} from '@common-types/index';

export interface TextFieldProps {
  name?: string;
  defaultValue?: string | number;
  disabled?: boolean;
  height?: string;
  width?: string;
  readonly?: boolean;
  required?: boolean;
  type?: TypeVariables;
  placeholder: string;
  iconUrl?: string;
  iconHeight?: string;
  iconWidth?: string;
  variant?: TextFieldVariants;
  label?: string;
  onInputChange?: (value: string | number, fieldName?: string) => void;
}

const TextField = forwardRef<HTMLInputElement, TextFieldProps>(
  (
    {
      name,
      defaultValue = '',
      disabled = false,
      height,
      label,
      width,
      type = TypeVariables.Text,
      readonly = false,
      required = false,
      placeholder,
      iconUrl,
      variant = TextFieldVariants.Standard,
      onInputChange
    },
    ref = null
  ) => {
    const [value, setValue] = useState<string | number>(defaultValue);

    const handleChange = useCallback((event: ChangeEvent) => {
      const newValue = (event.currentTarget as HTMLInputElement).value;
      setValue(newValue);

      onInputChange && onInputChange(newValue);
    }, []);

    return (
      <div className='field-wrapper'>
        {label && <label htmlFor={name}>{label}: </label>}
        <div className={'input-wrapper'}>
          {iconUrl && (
            <div className={'icon-left'}>
              <Image
                alt={'icon'}
                imageUrl={iconUrl}
                variant={ImageVariants.Icon}
              />
            </div>
          )}
          <input
            id={name}
            name={name}
            className={`field field-${variant}`}
            type={type}
            style={{ width, height }}
            placeholder={placeholder}
            value={value}
            disabled={disabled}
            readOnly={readonly}
            required={required}
            ref={ref}
            onChange={handleChange}
          />
        </div>
      </div>
    );
  }
);

TextField.displayName = 'TextField';

export default memo(TextField);
