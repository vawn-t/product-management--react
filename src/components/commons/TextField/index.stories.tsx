// Libraries
import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

// Components
import TextField from './index';

// Styles
import '@assets/styles/reset.css';
import '@assets/styles/App.css';
import '@assets/styles/reset.css';
import '@assets/styles/variables.css';

// Icons
import { searchIcon } from '@assets/index';

// Types
import { TextFieldVariants } from '@common-types/index';

export default {
  title: 'Components/TextField',
  component: TextField
} as ComponentMeta<typeof TextField>;

const Template: ComponentStory<typeof TextField> = (args) => (
  <TextField {...args} />
);

export const Standard = Template.bind({});
Standard.args = {
  label: 'Text field',
  placeholder: 'Placeholder',
  name: 'standard form'
};

export const Outline = Template.bind({});
Outline.args = {
  label: 'Text field',
  placeholder: 'Placeholder',
  variant: TextFieldVariants.Outline
};

export const Error = Template.bind({});
Error.args = {
  label: 'Text field',
  placeholder: 'Placeholder',
  variant: TextFieldVariants.Error
};

export const Disabled = Template.bind({});
Disabled.args = {
  label: 'Text field',
  placeholder: 'Placeholder',
  disabled: true
};

export const Readonly = Template.bind({});
Readonly.args = {
  placeholder: 'Placeholder',
  readonly: true
};

export const IconField = Template.bind({});
IconField.args = {
  placeholder: 'Placeholder',
  iconUrl: searchIcon,
  height: '4.25rem'
};
