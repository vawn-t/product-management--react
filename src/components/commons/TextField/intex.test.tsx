// Libraries
import React from 'react';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import { create } from 'react-test-renderer';
import '@testing-library/jest-dom';

// Components
import TextField, { TextFieldProps } from './index';

describe('TextField component', () => {
  const defaultProps: TextFieldProps = {
    placeholder: 'Placeholder'
  };

  let container: HTMLElement;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const standard = create(<TextField {...defaultProps} />).toJSON();
    expect(standard).toMatchSnapshot();

    const outline = create(<TextField {...defaultProps} />).toJSON();
    expect(outline).toMatchSnapshot();

    const error = create(<TextField {...defaultProps} />).toJSON();
    expect(error).toMatchSnapshot();
  });

  test('should render disable input', () => {
    render(<TextField {...defaultProps} disabled />, { container });

    const input = container.querySelector<HTMLInputElement>('input');

    expect(input).toHaveAttribute('disabled');
  });

  test('should render readonly input', () => {
    render(<TextField {...defaultProps} readonly />, { container });

    const input = container.querySelector<HTMLInputElement>('input');

    expect(input).toHaveAttribute('readonly');
  });

  test('should render TextField with label', () => {
    const label = 'Label';

    render(<TextField {...defaultProps} label={label} />, { container });

    const labelElement: HTMLLabelElement =
      container.getElementsByTagName('label')[0];

    expect(labelElement).toHaveTextContent(label);
  });

  test('should render image if we pass iconUrl', () => {
    const iconUrl = 'ICON_URL';

    render(<TextField {...defaultProps} iconUrl={iconUrl} />, { container });

    const image: HTMLImageElement = container.getElementsByTagName('img')[0];

    expect(image).toBeTruthy();
  });

  test('should render correct value', () => {
    const defaultValue = 'Test';

    render(<TextField {...defaultProps} defaultValue={defaultValue} />);

    const input: HTMLInputElement = screen.getByDisplayValue(defaultValue);

    expect(input.value).toMatch(defaultValue);
  });

  test('should be changed', () => {
    const myMock = jest.fn();
    const displayText = 'Test';

    render(
      <TextField
        {...defaultProps}
        onInputChange={myMock}
        defaultValue={displayText}
      />
    );

    const input: HTMLInputElement = screen.getByDisplayValue(displayText);

    fireEvent.change(input, { target: { value: 'Test1' } });

    expect(myMock).toHaveBeenCalled();
  });
});
