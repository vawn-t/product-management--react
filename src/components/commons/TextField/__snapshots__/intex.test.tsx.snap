// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField component should render correctly 1`] = `
<div
  className="field-wrapper"
>
  <div
    className="input-wrapper"
  >
    <input
      className="field field-standard"
      disabled={false}
      onChange={[Function]}
      placeholder="Placeholder"
      readOnly={false}
      required={false}
      style={
        Object {
          "height": undefined,
          "width": undefined,
        }
      }
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`TextField component should render correctly 2`] = `
<div
  className="field-wrapper"
>
  <div
    className="input-wrapper"
  >
    <input
      className="field field-standard"
      disabled={false}
      onChange={[Function]}
      placeholder="Placeholder"
      readOnly={false}
      required={false}
      style={
        Object {
          "height": undefined,
          "width": undefined,
        }
      }
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`TextField component should render correctly 3`] = `
<div
  className="field-wrapper"
>
  <div
    className="input-wrapper"
  >
    <input
      className="field field-standard"
      disabled={false}
      onChange={[Function]}
      placeholder="Placeholder"
      readOnly={false}
      required={false}
      style={
        Object {
          "height": undefined,
          "width": undefined,
        }
      }
      type="text"
      value=""
    />
  </div>
</div>
`;
